### ISSUE TITLE
[Idea] Transcription/translation via neural net

### ISSUE THREAD

**Issue by mRcSchwering** [mRcSchwering/magic-soup#2]
One big bottleneck currently is the fact that I am processing the genomes as strings. I convert them from strings to lists of proteins, and only then into cell params (matrices on GPU). But I could think of translating the genome to tokens first, and then immediately moving to GPU (or as a second step even only working on tokens and not string genomes). Some involved steps with pytorch:

1. Strings to tokens on CPU, then move to GPU [Example for simple mapping](https://stackoverflow.com/questions/65565461/how-to-map-element-in-pytorch-tensor-to-id)
2. have custom NN which translates tokens to vectors which contain domain information
3. translate domain vectors to cell params. Here, I probably need either a RNN or some clever aggregation [Exmple groupby using pytorch](https://discuss.pytorch.org/t/groupby-aggregate-mean-in-pytorch/45335)

For **2** I could think of having a padded CDS token sequence and then use a ConvNN, but:

- How do I make sure they e.g. don't overlap. Only the first one should be used in case of overlap (maybe RNN?)
- How to easily setup these NNs? Eventually the user should be able to just define some high level settings and not worry about what weights some RNN would need in order to do this arcane transformation.

Another thing: For mere computation this all might work, but eventually I also want to take a look at the cells. And then it would be better to actually have a list of proteins. But I can't reproduce a list of proteins from the cell params (they are degenerated). So, I need to still have the "classic" method of deriving proteins from the genome. But then I need to make sure the "new" and "classic" method both align.

**Comment by mRcSchwering** [comment]
I tried this for a week now with some success:

- I can use matmuls to map token sequences to feature vectors
- to create the mappings from any user input reliably I have to one-hot encode the tokens
- for extracting domains that don't overlap, I can use an RNN
- but again I need to do this on one-hot-encoded tokens
- the matrices become quite big (cells x proteins x domains x domain length x one-hot-encoding)
- even with small simulations RAM is already >10GB and it's overall slower (at least on CPU) than before
- very inefficient, most of the matrices are 99.9% 0s (one-hot-encoding, masks for either domain type, masks for implementing domain regions)

My middle-ground solution for now is to do the domain extraction from CDSs with mapping to tokens on the CPU.
These tokens now already represent a feature vector and can be multiplied with (dense) matrices that map these tokens to cell parameters (vectors for N, A and scalars for Km, Vmax, orient). To make this work every domain definition has to have the same structure, so that it can be multiplied with any mapping matrix (before I made a distinction between domain types). Now, each domain is defined by a variable region that defines the domain type (and whether it is a domain at all) and a constant region that defines the domain specification. The domain specification has first 1 2-codon region, then 3 1-codon regions. The 2-codon region can be used for mapping vectors (4096 possibilities), the 1-codon regions can be used for mapping scalars (64 possibilities). This will be fine until somebody starts defining more than 4096 molecule species or reactions (in which case I would need to increase this).

This is all implemented in [0.2.0](https://github.com/mRcSchwering/magic-soup/releases/tag/0.2.0). In roughly halved the processing time for setting cell parameters in small simulations. The biggest bottleneck is still the CPU part but it makes up a smaller amount now.
